### PR TITLE
Minimize dirtying of `swift build` relative to sourcekit-lsp invocations.

### PR DESCRIPTION
**Pull request by fischman-bcny**
### Motivation:

On Windows, building on the command line using `swift build` and also using VSCode causes `sourcekit-lsp` to modify the `.build` tree, making subsequent `swift build`s do nontrivial work instead of being null builds.
With this PR in SwiftPM that the sourcekit-lsp is built against, this extra work is no longer done 🎉 

### Modifications:

- replaced a couple of `writeFileContents` calls with `writeIfChanged` to avoid updating modification times, triggering unnecessary work in future builds.
- capitalize drive letters on Windows when writing `resource_bundle_accessor.swift` files since VSCode (and therefore vscode-swift) send lowercase drive letters to `sourcekit-lsp`, again dirtying the output tree unnecessarily. (vscode seems unlikely to change to use capital letters given e.g. https://github.com/microsoft/vscode/issues/42159#issuecomment-360791384)

### Result:
The latter build in the command-line below is near instant and does no work, as expected, instead of rebuilding a whole lot of stuff: `swift build && code . && sleep 30 && swift build`

### CODE REVIEW

File: Sources/Workspace/Workspace+State.swift [hunk 1]
```diff
@@ -254,7 +254,7 @@ extension WorkspaceStateStorage {
                         let version = try container.decode(String.self, forKey: .version)
                         return try self.init(underlying: .registryDownload(version: TSCUtility.Version(versionString: version)))
                     case "edited":
-                        let path = try container.decode(AbsolutePath?.self, forKey: .path)
+                        let path = try container.decode(AbsolutePath.self, forKey: .path)
```
Comment (tomerd) [inline Sources/Workspace/Workspace+State.swift:257]: Are these changes related?

File: Sources/Build/BuildDescription/SwiftTargetBuildDescription.swift [hunk 2]
```diff
@@ -348,7 +348,7 @@ public final class SwiftTargetBuildDescription {
             extension Foundation.Bundle {
                 static let module: Bundle = {
                     let mainPath = \(mainPathSubstitution)
-                    let buildPath = "\(bundlePath.pathString.asSwiftStringLiteralConstant)"
+                    let buildPath = "\(self.fileSystem.capitalizeDriveLetterIfOnWindows(bundlePath.pathString.asSwiftStringLiteralConstant))"
```
Comment (tomerd) [inline Sources/Build/BuildDescription/SwiftTargetBuildDescription.swift:351]: Is this change related?

Comment (tomerd) [reply Sources/Build/BuildDescription/SwiftTargetBuildDescription.swift:351]: cc @compnerd for opinions on this

Comment (tomerd) [reply Sources/Build/BuildDescription/SwiftTargetBuildDescription.swift:351]: also this may be better expressed as an extension of AbsolutePath instead of en extension of FileSystem

File: Sources/Workspace/Workspace+State.swift [hunk 3]
```diff
@@ -135,7 +135,7 @@ fileprivate struct WorkspaceStateStorage {
             let storage = V6(dependencies: dependencies, artifacts: artifacts)
 
             let data = try self.encoder.encode(storage)
-            try self.fileSystem.writeFileContents(self.path, data: data)
+            try self.fileSystem.writeIfChanged(path: self.path, bytes: .init(data))
```
Comment (tomerd) [inline Sources/Workspace/Workspace+State.swift:138]: consider an wrapper for writeIfChanged that takes data for future similar use cases

**Comment by tomerd** [comment]
thanks @fischman-bcny, some follow up ideas and questions

File: Sources/Workspace/Workspace+State.swift (hunk 1 above)
Comment (fischman-bcny) [reply Sources/Workspace/Workspace+State.swift:257]: Unrelated but required for this package to build in the first place with a tip-of-tree (or near-to-that) swiftc; see PR descriptions: 
> unrelated fixes to appease snapshot swiftc where AbsolutePath?.self is no longer valid Swift.

File: Sources/Workspace/Workspace+State.swift (hunk 3 above)
Comment (fischman-bcny) [reply Sources/Workspace/Workspace+State.swift]: Done.

File: Sources/Build/BuildDescription/SwiftTargetBuildDescription.swift (hunk 2 above)
Comment (fischman-bcny) [reply Sources/Build/BuildDescription/SwiftTargetBuildDescription.swift:351]: > Is this change related?

Yes, see PR description's 2nd bullet (VSCode passes "c:\..." but `swift build` gets here with "C:\..." so we need to capitalize the `c` to `C` to avoid spurious changes to the emitted files)

> also this may be better expressed as an extension of AbsolutePath instead of en extension of FileSystem

No preference on my end.

**Review by fischman-bcny — COMMENTED** [review]
Thanks @tomerd for the (very!) quick turnaround; please see responses inline.

File: Sources/Workspace/Workspace+State.swift (hunk 1 above)
Comment (tomerd) [reply Sources/Workspace/Workspace+State.swift:257]: I dont think we have any issue building SwiftPM with ToT. where have you seen this to be an issue?

cc @neonichu

Comment (fischman-bcny) [reply Sources/Workspace/Workspace+State.swift:257]: Sorry, I didn't note down the failure when I saw it, and now I can't reproduce it anymore. 

I'm fine reverting this aspect of the PR, but is there a context in which the `?` is adding value?

Comment (tomerd) [reply Sources/Workspace/Workspace+State.swift:257]: it is saying that the value is optional so if the serialized formal (in this case I think it is JSON) does not have it its fine. this is important for backwards compatibility

Comment (fischman-bcny) [reply Sources/Workspace/Workspace+State.swift]: Oh, I see. Thanks for catching this. Reverted.

File: Sources/Build/BuildDescription/SwiftTargetBuildDescription.swift (hunk 2 above)
Comment (compnerd) [reply Sources/Build/BuildDescription/SwiftTargetBuildDescription.swift:351]: @tomerd hmm, I think that the ideal thing would be to ensure that we always normalise the drive letter spelling.  I know that node.js will case convert to lower case, but that changes the path technically and thus causes rebuilds.  Translating to uppercase seems like the right thing to do IMO - that is the native windows spelling and the Windows APIs spell it that way.

File: Sources/Basics/FileSystem/FileSystem+Extensions.swift [hunk 4]
```diff
@@ -198,6 +198,18 @@ extension FileSystem {
...
+    public func capitalizeDriveLetterIfOnWindows(_ path: String) -> String {
+        #if !os(Windows)
+        return path
+        #endif
+        guard path.count > 2,
+            path.prefix(2).last == ":",
+            let drive = path.first,
+            drive >= "a" && drive <= "z"
+        else { return path }
+        return drive.uppercased() + path.suffix(path.count - 1)
```
Comment (compnerd) [inline Sources/Basics/FileSystem/FileSystem+Extensions.swift:211]: I think that I would rather see this spelt:

```swift
var iter = path.makeIterator()
guard iter.next().isLetter.isUppercase && iter.next() == ":" else { return path }
return "\(path.first.uppercased())\(path.dropFirst())"
```

I do think that @tomerd has a point on this being part of `AbsolutePath`.  I think that we want this normalised in all cases.

File: Sources/Build/BuildDescription/SwiftTargetBuildDescription.swift (hunk 2 above)
Comment (neonichu) [reply Sources/Build/BuildDescription/SwiftTargetBuildDescription.swift:351]: I agree with @compnerd, this feels like something `pathString` or `asSwiftStringLiteralConstant` should be doing vs. an extra manual step.

Comment (tomerd) [reply Sources/Build/BuildDescription/SwiftTargetBuildDescription.swift:351]: I would be fine with that approach if that is the canonical representation on windows

Comment (tomerd) [reply Sources/Build/BuildDescription/SwiftTargetBuildDescription.swift:351]: we should consider splitting this into 2 PRs. this one can focus on the write-if-changed, and a second one can focus on canonicalizing / normalizing the drive letter in AbsolutePath on windows

File: Sources/Basics/FileSystem/FileSystem+Extensions.swift (hunk 4 above)
Comment (compnerd) [reply Sources/Basics/FileSystem/FileSystem+Extensions.swift:211]: "in all cases" means in all cases.  We want this to happen always when constructing a path, not just when constructing the contents for the SPM generated bundle.

Sinking this into t-s-c should effectively accomplish that I believe: https://github.com/apple/swift-tools-support-core/pull/422.
